### PR TITLE
server side render JSX

### PR DESCRIPTION
**Pull request by will-weiss**
- Uses JSX for templating
- Organize stylesheets
- Use JSDOM for testing, loading CSS and JS from the same server

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,15 +16,20 @@
     "@sentry/node": "5.27.3",
     "@sentry/tracing": "5.27.3",
     "axios": "0.21.0",
+    "babel-preset-es2015": "6.24.1",
```
Comment (shachar-langer) [inline package.json:19]: Shouldn't `babel` and `jsdom` be under the `devDependencies`?

File: src/server.ts [hunk 2]
```diff
@@ -25,6 +33,8 @@ export function createServer(withRouter?: (router: Router) => Router): Koa {
     })
   }
 
+  react(server, { views: process.cwd() + '/views' })
```
Comment (shachar-langer) [inline src/server.ts]: We can theoretically run the server from any folder. Let's use `__dirname` and `path.join` for Windows users.

File: public/auth-success.html [hunk 3]
```diff
@@ -1 +0,0 @@
-<p>You did it!</p>
```
Comment (shachar-langer) [inline public/auth-success.html:1]: Should we remove the `successRedirect` from [handlers.ts](https://github.com/morehumaninternet/roar-server/blob/main/src/handlers.ts#L87)?

File: public/styles/welcome.css [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+main {
+  flex-grow: 1;
+  background: linear-gradient(var(--human-pink-light), white);
+
+  display: flex;
+  place-items: center;
+}
+
+/* https://www.joshwcomeau.com/css/full-bleed/ */
```
Comment (shachar-langer) [inline public/styles/welcome.css:16]: I haven't read this article yet, but it looks like we're not using some of the selectors here.

File: src/__test__/integration/landing-page.test.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+describe('landing page', () => {
+  const mocks = createMocks()
+
+  let window: DOMWindow
+  before(async () => window = await mocks.getWindow('/'))
+
+  it('has a header', () => {
+    const style = window.getComputedStyle(window.document.querySelector('header')!)
+    expect(style.paddingTop).to.equal('10px')
```
Comment (shachar-langer) [inline src/__test__/integration/landing-page.test.ts:15]: That's a very specific test, but I guess it's ok. Maybe we can take a screenshot of the header and see it hasn't changed?

File: src/__test__/mocks/getWindow.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+export async function getWindow(agent: request.SuperTest<request.Test>, path: string): Promise<DOMWindow> {
+  const response = await agent.get(path)
+
+  const { window } = new JSDOM(response.text, {
+    url: fakeUrl,
+    pretendToBeVisual: true,
+    runScripts: 'dangerously',
+    resources: Object.assign(new ResourceLoader(), {
```
Comment (shachar-langer) [inline src/__test__/mocks/getWindow.ts:36]: I read [this](https://github.com/jsdom/jsdom#loading-subresources) section from JSDom docs and I don't understand why we can use `resources: "usable"`. Can you please explain?

File: src/router.ts [hunk 7]
```diff
@@ -2,9 +2,28 @@ import { identity } from 'lodash'
...
 
+declare module 'koa-router' {
+  export interface IRouterContext {
+    render(view: string, data?: any): any
+  }
+}
+
+// Sychronously yield all files in the given directory, searching recursively
+function * files(dir: string): IterableIterator<ParsedPath & { full: string }> {
+  for (const name of readdirSync(process.cwd() + dir)) {
```
Comment (shachar-langer) [inline src/router.ts]: Using `process.cwd()` and the `+` sign is risky. Let's use `__dirname` and `path.join`

File: src/router.ts [hunk 8]
```diff
@@ -18,16 +37,14 @@ export function createRouter(withRouter: (router: Router) => Router = identity):
...
 
-  // tslint:disable-next-line:no-expression-statement
-  readdirSync(process.cwd() + '/public').forEach(fileName => {
-    const file = parse(fileName)
-    const route = file.ext === '.html' ? file.name : fileName
-    return router.get('/' + route, ctx => send(ctx, `/public/${fileName}`))
-  })
+  for (const file of files('/public')) {
+    const route = file.ext === '.html' ? file.name : file.base
+    router.get('/' + route, ctx => send(ctx, file.full)) // tslint:disable-line:no-expression-statement
```
Comment (shachar-langer) [inline src/router.ts:47]: Can't we use `ctx.render` here as well?

File: views/index.jsx [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            <svg viewBox="0 0 418 355">
+              <path d="M38 0C17.005 0 0 15.8862 0 35.5V248.5C0 268.114 17.005 284 38 284H171V319.5H133V355H285V319.5H247V284H380C400.995 284 418 268.114 418 248.5V35.5C418 15.8862 400.995 0 380 0H38ZM38 35.5H380V248.5H38V35.5ZM190 71V159.75H228V71H190ZM190 177.5V213H228V177.5H190Z" fill="#FB945B"/>
+            </svg>
+          }
+        />
+        <Step
+          title="2. Take a screenshot"
+          description="Open the Roar! extension to automatically capture a screenshot of the issue."
+          icon={
+            <svg viewBox="0 0 30 34">
```
Comment (shachar-langer) [inline views/index.jsx]: All these inline `SVG`s make it really hard to read the code. Is it possible to split them to separate files?

File: package.json (hunk 1 above)
Comment (will-weiss) [reply package.json:19]: jsdom yes, but babel is used by the runtime.

File: src/server.ts (hunk 2 above)
Comment (will-weiss) [reply src/server.ts]: good point, will do

File: public/auth-success.html (hunk 3 above)
Comment (will-weiss) [reply public/auth-success.html:1]: We're still detecting the user being redirected to this blank page by the extension. Not sure if we don't need any success redirect

File: public/styles/welcome.css (hunk 4 above)
Comment (will-weiss) [reply public/styles/welcome.css:16]: Yeah, I'll clean this up when doing #48

File: src/__test__/integration/landing-page.test.ts (hunk 5 above)
Comment (will-weiss) [reply src/__test__/integration/landing-page.test.ts:15]: Yeah, just wanted to prove that the styles are working

File: src/__test__/mocks/getWindow.ts (hunk 6 above)
Comment (will-weiss) [reply src/__test__/mocks/getWindow.ts:36]: We only want to fetch specific resources, namely the ones from our own server. Perhaps we want to do this over the network, but using the `agent` worked

File: src/router.ts (hunk 8 above)
Comment (will-weiss) [reply src/router.ts:47]: These are mostly for the css and favicon files. The auth-success page is weird, but I think we can leave it in for now

File: package.json (hunk 1 above)
Comment (shachar-langer) [reply package.json:19]: oh yeah, we are rendering the views on the server. I wonder if we can do it statically while building the server.

**Review by shachar-langer — APPROVED** [review]

**Pull request merged**